scripts/e2e_test.sh

```diff
@@ -1,36 +1,197 @@
+done
+
+# Execute fuzzing process
+echo "Starting fuzzing process (timeout: $MAKE_TIMEOUT)..."
+mkdir -p "$FUZZ_RESULTS_PATH"
+MAKE_LOG="$FUZZ_RESULTS_PATH/make_run.log"
+
+# Run make run under timeout, capturing stdout+stderr into MAKE_LOG.
+timeout -s INT --preserve-status "$MAKE_TIMEOUT" make run ARGS="$ARGS" 2>&1 | tee "$MAKE_LOG"
+status=${PIPESTATUS[0]}
+
+# Handle exit codes:
+#   130 → timeout sent SIGINT; treat as expected termination
+#   any other non-zero → unexpected error
+if [[ $status -ne 130 ]]; then
+  echo "❌ Fuzzing exited with unexpected error (status: $status)."
+  exit "$status"
 fi
 
-# Cleanup: Delete the $HOME/fuzz_results directory
-rm -rf "$HOME/fuzz_results"
\ No newline at end of file
+# List of required patterns to check in the log
+readonly REQUIRED_PATTERNS=(
+  'Cycle duration complete; initiating cleanup.'
+  'Fuzzing completed successfully'
+  'gathering baseline coverage'
+  'Shutdown initiated during fuzzing cycle; performing final cleanup.'
+)
+
+# Verify that worker logs contain expected entries
+echo "Verifying worker log entries in $MAKE_LOG..."
+for pattern in "${REQUIRED_PATTERNS[@]}"; do
+  if ! grep -q -- "$pattern" "$MAKE_LOG"; then
+    echo "❌ ERROR: Missing expected log entry: $pattern"
+    exit 1
+  fi
+done
+
+# Capture final corpus state
+echo "Recording final corpus state..."
+for target in "${FUZZ_TARGETS[@]}"; do
+  IFS=':' read -r pkg func <<<"$target"
+  echo "  - $pkg/$func"
+  final_input_counts["$target"]=$(count_corpus_inputs "$pkg" "$func")
+  final_coverage_metrics["$target"]=$(measure_fuzz_coverage "$pkg" "$func")
+done
+
+# Validate corpus growth
+echo "Validating corpus growth..."
+for target in "${FUZZ_TARGETS[@]}"; do
+  initial_count=${initial_input_counts["$target"]}
+  final_count=${final_input_counts["$target"]}
+
+  if [[ $final_count -lt $initial_count ]]; then
+    echo "❌ ERROR: $target regressed - inputs decreased from $initial_count to $final_count"
+    exit 1
+  fi
+done
+
+# Validate coverage metrics
+echo "Validating coverage metrics..."
+for target in "${FUZZ_TARGETS[@]}"; do
+  initial_cov=${initial_coverage_metrics["$target"]}
+  final_cov=${final_coverage_metrics["$target"]}
+
+  if [[ $final_cov -lt $initial_cov ]]; then
+    echo "❌ ERROR: $target coverage decreased from $initial_cov to $final_cov"
+    exit 1
+  fi
+done
+
+# Verify crash reports
+echo "Checking crash reports..."
+required_crashes=(
+  "$FUZZ_RESULTS_PATH/FuzzParseComplex_failure.log"
+  "$FUZZ_RESULTS_PATH/FuzzUnSafeReverseString_failure.log"
+)
+
+for crash_file in "${required_crashes[@]}"; do
+  if [[ ! -f "$crash_file" ]]; then
+    echo "❌ ERROR: Missing crash report: $crash_file"
+    exit 1
+  fi
+
+  if ! grep -q "go test fuzz v1" "$crash_file"; then
+    echo "❌ ERROR: Invalid crash report format in $crash_file"
+    exit 1
+  fi
+done
+
+# Cleanup resources
+echo "Cleaning up resources..."
+rm -rf "$PROJECT_DIR"
+rm -rf "$FUZZ_RESULTS_PATH"
```

Comment:
Note that this cleanup gets skipped when the test fails.  We could use a trap to ensure the cleanup always occurs.

But maybe some of the data would be helpful to debug the failure?  What do you think?